### PR TITLE
Fix stack overwrite

### PR DESCRIPTION
**Pull request by JuergenKosel**
This pull request fixes a stack overwrite, which is caused by bad return value handling
of Socket_writev() in src/Socket.c

### CODE REVIEW

File: src/Socket.c [hunk 1]
```diff
@@ -473,14 +476,16 @@ int Socket_putdatas(int socket, char* buf0, size_t buf0len, int count, char** bu
 	iovecs[0].iov_base = buf0;
 	iovecs[0].iov_len = (ULONG)buf0len;
 	frees1[0] = 1;
+	assert(count <= 5);
```
Comment (JuergenKosel) [inline src/Socket.c:479]: Here is another possible stack overwrite, if the caller sets count > 5 !

File: src/Socket.c [hunk 2]
```diff
@@ -747,6 +752,7 @@ int Socket_continueWrite(int socket)
 	}
 #endif
 
+	assert(5 >= pw->count);
```
Comment (JuergenKosel) [inline src/Socket.c:755]: Here is another possible stack overwrite, if the caller sets pw->count > 5 !

**Comment by icraggs** [comment]
Thanks Juergen.  Do you actually encounter this situation in real use - if so what are the circumstances?

**Comment by JuergenKosel** [comment]
Hello @icraggs 

I have seen this stack corruption in real use: The broker was disconnected from the LAN while the paho.mqtt.c client has a number of items to publish.

File: src/Socket.c [hunk 3]
```diff
@@ -425,15 +425,18 @@ int Socket_writev(int socket, iobuf* iovecs, int count, unsigned long* bytes)
...
 #else
 	*bytes = 0L;
+        assert(count > 0);
 	rc = writev(socket, iovecs, count);
-	if (rc == SOCKET_ERROR)
+	if (rc < 0)
 	{
 		int err = Socket_error("writev - putdatas", socket);
 		if (err == EWOULDBLOCK || err == EAGAIN)
 			rc = TCPSOCKET_INTERRUPTED;
```
Comment (JuergenKosel) [inline src/Socket.c:434]: In this line, the return value is set to -22.

File: src/Socket.c [hunk 4]
```diff
@@ -766,7 +772,8 @@ int Socket_continueWrite(int socket)
 		curbuflen += pw->iovecs[i].iov_len;
 	}
 
-	if ((rc = Socket_writev(socket, iovecs1, curbuf+1, &bytes)) != SOCKET_ERROR)
```
Comment (JuergenKosel) [inline src/Socket.c:769]: Here the return value -22 was not detected as failure.
So the random value in `bytes` was used. And this has lead to stack corruption in a following call of writev().

File: src/Socket.c (hunk 3 above)
Comment (icraggs) [reply src/Socket.c:434]: writev returns -1 on error, so this shouldn't change anything, right?

File: src/Socket.c (hunk 2 above)
Comment (icraggs) [reply src/Socket.c:755]: Again, this change will have no effect, that I can see.  (Apart from the assertion, but I don't think we want assertions in production code).

File: src/Socket.c (hunk 4 above)
Comment (icraggs) [reply src/Socket.c:769]: I don't think the value in bytes will be random - it is initialised to 0 in Socket_writev, and then not written to unless the return from writev was not -1.

**Review by icraggs — COMMENTED** [review]
The analysis doesn't look right to me.  I think we should start from the symptoms, and work from there.

File: src/Socket.c (hunk 3 above)
Comment (JuergenKosel) [reply src/Socket.c:434]: In line 433 the return value is changed to -22 (TCPSOCKET_INTERRUPTED), and this value is not handled correctly by the caller!

File: src/Socket.c (hunk 2 above)
Comment (JuergenKosel) [reply src/Socket.c:755]: Socket_writev() can return -22, and this was not handled here!

The assertions will be disabled, if compiled with macro NDEBUG defined. But for regression tests, this assertions may be useful. At least it is easier to debug a "crash" caused by an assertion, than some random behaviour caused by stack overwrite!

File: src/Socket.c [hunk 5]
```diff
@@ -425,15 +425,18 @@ int Socket_writev(int socket, iobuf* iovecs, int count, unsigned long* bytes)
 	}
 #else
 	*bytes = 0L;
```
Comment (JuergenKosel) [inline src/Socket.c:427]: I have over seen this line.

**Comment by JuergenKosel** [comment]
I have over seen, that bytes is set to 0 inside Socket_writev().
 So my modification prevent the segfault / stack corruption by accident. Because it prevent correct storing of the pending write.

File: src/Socket.c (hunk 2 above)
Comment (icraggs) [reply src/Socket.c:755]: I think it is correct, or at least my intention, that the execution path goes into the subsequent code block if the rc is -22 (interrupted).

**Pull request closed without merging**